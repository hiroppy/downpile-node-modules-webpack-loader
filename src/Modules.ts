import { sep } from 'path';
import { Options } from 'acorn';
import { flattenPackageLockDeps, PackageLock } from 'flatten-package-lock-deps';

export class Modules {
  modulesMap: ReturnType<typeof flattenPackageLockDeps>;
  cacheList: {
    targetVersion: number;
    modules: {
      [key in string]?: {
        isScopedVersion: 'yes' | 'no';
      };
    };
  };

  constructor(lockFile: PackageLock, version: Options['ecmaVersion'] = 5) {
    this.modulesMap = flattenPackageLockDeps(lockFile, { ignoreDev: true });
    // readFile cache
    this.cacheList = {
      targetVersion: version,
      modules: {},
    };
  }

  // node_modules/<target>
  // node_modules/yyyy/node_modules/<target>
  parseModuleInfoFromContext(context: string) {
    const modulePathArr = context.split(sep);
    let parent = '';
    let name = '';

    // depth === the number of node_modules
    let depth = -1;

    for (let i = 0; i < modulePathArr.length; i++) {
      const p = modulePathArr[i];

      if (p === 'node_modules') {
        ++depth;

        if (depth >= 1) {
          parent = name;
        }

        name = modulePathArr[i + 1];
      }
    }

    return {
      parent,
      name,
      depth,
    };
  }

  getModuleInfoFromContext(context: string) {
    const { name, depth, parent } = this.parseModuleInfoFromContext(context);
    const { integrity, version } =
      Object.values(this.modulesMap).find((m) => {
        if (m.name === name) {
          // node_modules/xxx, it means this module is written in package.json
          if (depth === 0 && m.depth === 0) {
            return true;
          }
          // node_modules/xxx/node_modules/yyy
          if (m.parents.includes(parent)) {
            return true;
          }
        }
        return false;
      }) || {};

    if (!integrity) {
      return null;
    }

    return {
      dirName: `${name}-${version}`,
      name,
      depth,
      version,
      integrity,
    };
  }

  getModuleCache() {
    return this.cacheList;
  }

  getModuleFromCache(moduleName: string) {
    return this.cacheList.modules[moduleName];
  }

  pushToCache(name: string, data: { isScopedVersion: 'yes' | 'no' }) {
    this.cacheList.modules[name] = data;
  }
}
