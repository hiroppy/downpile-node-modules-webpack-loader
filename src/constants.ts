export const loaderName = 'downpile-node-modules-webpack-loader';
export const defaultVersion = 5;
