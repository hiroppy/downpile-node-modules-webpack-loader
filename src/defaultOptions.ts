export const defaultOptions = {
  ignoredLibraries: [],
  polyfill: false,
  ecmaVersion: 5,
  // for CI
  // outputModulesList:
  // readModulesList
};
