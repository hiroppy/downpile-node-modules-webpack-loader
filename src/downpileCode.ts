import * as acorn from 'acorn';
import { transformAsync } from '@babel/core';
import presetEnv from '@babel/preset-env';

export async function downpileCode(
  src: string,
  version = 5,
  options: {
    polyfill?: boolean;
  } = {}
): Promise<{
  src: string;
  isScopedVersion: 'yes' | 'no';
}> {
  try {
    acorn.parse(src, { ecmaVersion: version as acorn.Options['ecmaVersion'] });

    return {
      src,
      isScopedVersion: 'yes',
    };
  } catch (err) {
    // @ts-ignore
    const { code } = await transformAsync(src, {
      presets: [
        [
          presetEnv,
          {
            modules: false,
            useBuiltIns: options.polyfill ? 'usage' : false,
          },
        ],
      ],
    });

    return {
      src: code,
      isScopedVersion: 'no',
    };
  }
}
