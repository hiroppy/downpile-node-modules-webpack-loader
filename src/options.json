{
  "additionalProperties": false,
  "type": "object",
  "properties": {
    "ignoredLibraries": {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "polyfill": {
      "type": "boolean"
    },
    "ecmaVersion": {
      "anyOf": [
        {
          "type": "string"
        },
        {
          "type": "number"
        }
      ]
    }
  }
}
