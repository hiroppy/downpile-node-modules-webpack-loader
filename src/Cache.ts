import { promises } from 'fs';
import { join } from 'path';
import del from 'del';
import findCacheDir from 'find-cache-dir';
import { loaderName } from './constants';

const { readFile, writeFile, mkdir } = promises;

export class Cache {
  name: string;
  cacheDirPath: string;

  constructor() {
    this.name = loaderName;
    this.cacheDirPath = this.getCacheDirPath();
  }

  getCacheDirPath() {
    return findCacheDir({ name: this.name }) as string;
  }

  createFilename(internalPath: string) {
    if (!internalPath.startsWith('.')) {
      return internalPath;
    }

    const fileName = internalPath.replace(/\./g, '_').replace(/\//g, '.');

    return fileName;
  }

  getFilePath(moduleName: string, fileName: string) {
    return join(this.cacheDirPath, moduleName, fileName);
  }

  async createCacheDir() {
    await mkdir(this.cacheDirPath, { recursive: true });
  }

  async deleteCacheDir() {
    await del(this.cacheDirPath);
  }

  async createModuleDir(name: string) {
    await mkdir(join(this.cacheDirPath, name), { recursive: true });
  }

  async writeData(moduleName: string, fileName: string, content: string) {
    await writeFile(this.getFilePath(moduleName, fileName), content);
  }

  async readData(moduleName: string, fileName: string) {
    return await readFile(this.getFilePath(moduleName, fileName), 'utf-8');
  }
}
