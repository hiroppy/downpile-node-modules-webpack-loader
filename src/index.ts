import { join } from 'path';
import { loader } from 'webpack';
import { getOptions } from 'loader-utils';
import validateOptions from 'schema-utils';
// const exitHook = require('exit-hook');
import { loaderName, defaultVersion } from './constants';
import { defaultOptions } from './defaultOptions';
import { Modules } from './Modules';
import { Cache } from './Cache';
import { downpileCode } from './downpileCode';
import schema from './options.json';

let initialFlag = false;
const lockFile = require(join(process.cwd(), 'package-lock.json'));
const modules = new Modules(lockFile);
const cache = new Cache();

// TODO: if build finishes, we need to output moduleList
// TODO: emitError
export default async function downpileLoader(this: loader.LoaderContext, src: string) {
  const custom = getOptions(this);
  const options: NonNullable<typeof defaultOptions> = <const>{
    ...defaultOptions,
    ...custom,
  };
  const cb = this.async()!;

  if (!initialFlag) {
    // TODO: need to check the type
    validateOptions(schema as any, options, {
      name: loaderName,
    });
    await cache.createCacheDir();

    // read modulesList from cache
    // versionが違ったら更新する、Modulesの初期化
    initialFlag = true;
  }

  const fileName = cache.createFilename(this._module.rawRequest);
  const data = modules.getModuleInfoFromContext(this._module.context);

  if (data === null) {
    cb(null, src);
    return;
  }

  const moduleFromCache = modules.getModuleFromCache(data.dirName);
  let resSrc = '';

  if (!moduleFromCache) {
    const { src: moduleCode, isScopedVersion } = await downpileCode(src, defaultVersion, options);

    if (data.version !== undefined) {
      if (isScopedVersion === 'no') {
        await cache.createModuleDir(data.dirName);
        await cache.writeData(data.dirName, fileName, moduleCode);
      }
      modules.pushToCache(data.dirName, { isScopedVersion });
    }

    resSrc = moduleCode;
  } else {
    // return src that is via webpack
    if (moduleFromCache.isScopedVersion === 'yes') {
      resSrc = src;
    } else {
      try {
        resSrc = await cache.readData(data.dirName, fileName);
      } catch (e) {
        // this loader has already known the module which should be downgraded but its file in the module doesn't exist at .cache.
        const { src: moduleCode } = await downpileCode(src, defaultVersion, options);

        await cache.writeData(data.dirName, fileName, moduleCode);
        resSrc = moduleCode;
      }
    }
  }

  cb(null, resSrc);
}

// TODO: write modulesList
// exitHook(() => {
//   console.log('Exiting');
// });
