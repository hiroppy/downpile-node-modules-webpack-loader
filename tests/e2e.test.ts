import { join } from 'path';
import process from 'process';
import del from 'del';
import { checkResult } from './helpers/checkResult';
import { getFilesList } from './helpers/getFilesList';
import { loaderName } from '../src/constants';

describe('e2e', () => {
  describe('simple', () => {
    const { base, getLoader } = require('./fixtures/simple/webpack.config');
    const cacheDir = join(__dirname, `./fixtures/simple/node_modules/.cache/${loaderName}`);

    afterEach(async () => {
      await del(cacheDir);
    });

    beforeAll(() => {
      jest.spyOn(process, 'cwd').mockReturnValue(join(__dirname, './fixtures/simple'));
    });

    it('es5', async () => {
      // check es5 or not
      await expect(checkResult(base, 5)).rejects.toBe("The keyword 'const' is reserved (1:928)");
      // add this loader
      await expect(checkResult({ ...base, ...getLoader() }, 5)).resolves.toBe('ok');

      expect(getFilesList(cacheDir)).toMatchSnapshot();
    });
  });

  describe.skip('nested', () => {
    const { base, getLoader } = require('./fixtures/nested/webpack.config');
    const cacheDir = join(__dirname, `./fixtures/nested/node_modules/.cache/${loaderName}`);

    beforeAll(() => {
      jest.spyOn(process, 'cwd').mockReturnValue(join(__dirname, './fixtures/nested'));
    });

    afterEach(async () => {
      await del(cacheDir);
    });

    it('es5', async () => {
      // add this loader
      await expect(checkResult({ ...base, ...getLoader() }, 5)).resolves.toBe('ok');

      expect(getFilesList(cacheDir)).toMatchSnapshot();
    });
  });
});
