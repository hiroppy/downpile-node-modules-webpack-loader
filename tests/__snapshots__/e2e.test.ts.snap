// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e nested es5 1`] = `
Object {
  "react-lifecycles-compat-3.0.4": Array [
    "react-lifecycles-compat",
  ],
}
`;

exports[`e2e simple es5 1`] = `
Object {
  "ansi-styles-4.2.1": Array [
    "ansi-styles",
  ],
  "chalk-4.0.0": Array [
    "_.templates",
    "_.util",
    "chalk",
  ],
  "color-convert-2.0.1": Array [
    "_.conversions",
    "_.route",
    "color-convert",
  ],
}
`;
